File: src/olcw78 (die yoong)/jsAlgorithmStudy/week05/24_신고_결과_받기.go
package week05

import (
	"fmt"
	"strings"
)

type reportRecord struct {
	reportedBy    map[string]string
	reportedCount int
}

// 각 유저는 한 번에 한 명의 유저를 신고할 수 있습니다.
// - 신고 횟수에 제한은 없다. 서로 다른 유저를 계속 신고 가능
// - 한 유저를 여러번 신고할 수 있지만 동일한 유저에 대한 신고 횟수는 1회로 처리됨.
// k번 이상 신고된 유저는 정지되며 해당 유저를 신고한 모든 유저에게 정지 사실을 메일로 발송
// - 유저가 신고한 모든 내용을 취합해 마지막에 한꺼번에 계시판 이용 정지를 시키면서 정지 메일을 발송
func solution24(id_list []string, report []string, k int) []int {
	idHt := map[string]int{}
	for i, id := range id_list {
		idHt[id] = i
	}
	history := map[string]reportRecord{}
	ret := make([]int, len(id_list))
	for _, r := range report {
		rsplit := strings.Split(r, " ")
		reporter, reported := rsplit[0], rsplit[1]

		if h, ok := history[reported]; !ok {
			history[reported] = reportRecord{
				reportedBy:    map[string]string{reporter: reporter},
				reportedCount: 1,
			}
		} else {
			newRecord := reportRecord{
				reportedBy:    map[string]string{reporter: reporter},
				reportedCount: h.reportedCount,
			}

			nextCnt := 0
			if _, ok2 := h.reportedBy[reporter]; !ok2 {
				nextCnt = h.reportedCount + 1
			} else {
			}

			for _, p := range h.reportedBy {
				newRecord.reportedBy[p] = p
			}
			newRecord.reportedCount = nextCnt
			history[reported] = newRecord
		}
	}

	for _, h := range history {
		for _, reporter := range h.reportedBy {
			if h.reportedCount >= k {
				ret[idHt[reporter]]++
			}
		}
	}

	return ret
}

func solution24_2(id_list []string, report []string, k int) []int {
	idMap := make(map[string]bool)
	for _, id := range id_list {
		idMap[id] = false
	}
	// 중복 신고 제거를 위한 맵
	reportSet := make(map[string]bool)
	// 신고 횟수를 카운트하는 맵
	reportCount := make(map[string]int)
	// 신고자별 신고한 사용자 맵
	userReport := make(map[string][]string)

	// 중복 제거 및 신고 처리
	for _, r := range report {
		if reportSet[r] {
			continue
		}
		reportSet[r] = true

		spl := strings.Split(r, " ")
		reporter, reported := spl[0], spl[1]
		reportCount[reported]++
		userReport[reporter] = append(userReport[reporter], reported)
	}

	// 결과 계산
	result := make([]int, len(id_list))
	for i, id := range id_list {
		for _, reported := range userReport[id] {
			if reportCount[reported] >= k {
				result[i]++
			}
		}
	}

	return result
}

func Run24() {
	r1 := solution24(
		[]string{"muzi", "frodo", "apeach", "neo"},
		[]string{
			"muzi frodo",
			"apeach frodo",
			"frodo neo",
			"muzi neo",
			"apeach muzi",
		},
		2,
	)
	fmt.Println(r1)
	// [2, 1, 1, 0]

	r2 := solution24(
		[]string{"con", "ryan"},
		[]string{"ryan con", "ryan con", "ryan con", "ryan con"},
		3,
	)
	fmt.Println(r2)
	// [0, 0]
}

File: src/olcw78 (die yoong)/jsAlgorithmStudy/week05/24_신고_결과_받기.js
/**
 * 시간 복잡도 (Time Complexity)
     * ID 매핑 생성
     * id_list를 순회: O(N), N은 id_list의 길이
     * 신고 기록 처리
     * report 배열을 순회: O(R), R은 report의 길이
     * 각 신고 처리에서 Set 연산: O(1)
     * 결과 계산
     * history 객체 순회: O(M), M은 신고된 유저 수 (최대 N)
     * 각 신고자 Set 순회: O(N) 최악의 경우
 * 총 시간 복잡도: O(N + R + N²) = O(N² + R)
 * 공간 복잡도 (Space Complexity)
     * idMap: O(N)
     * 모든 ID와 인덱스 저장
     * history: O(R)
     * 최악의 경우, 모든 신고 기록을 저장
     * Set 객체들이 포함됨
     * ret 배열: O(N)
     * 결과 배열
 * 총 공간 복잡도: O(N + R) 
 * @param {string[]} id_list 
 * @param {string[]} report 
 * @param {number} k 
 * @returns 
 */
function solution24(id_list, report, k) {
    // ID를 인덱스로 매핑하는 객체 생성
    const idMap = {};
    for (const [i, id] of id_list.entries()) {
        idMap[id] = i;
    }

    // 신고 기록을 저장할 객체
    /** @type {{ reportedBy: Set<string>, reportedCount: number }} */
    const history = {};
    const ret = new Array(id_list.length).fill(0);

    // 신고 기록 처리
    for (const r of report) {
        const [reporter, reported] = r.split(' ');

        if (!history[reported]) {
            history[reported] = {
                reportedBy: new Set([reporter]),
                reportedCount: 1
            };
        } else {
            history[reported].reportedBy.add(reporter);
            history[reported].reportedCount = history[reported].reportedBy.size;
        }
    }

    // 결과 계산
    Object.entries(history).forEach(([reported, data]) => {
        if (data.reportedCount >= k) {
            data.reportedBy.forEach(reporter => {
                ret[idMap[reporter]]++;
            });
        }
    });

    return ret;
}

// 테스트 실행
function run24() {
    const r1 = solution24(
        ["muzi", "frodo", "apeach", "neo"],
        [
            "muzi frodo",
            "apeach frodo",
            "frodo neo",
            "muzi neo",
            "apeach muzi"
        ],
        2
    );
    console.log(r1); // [2, 1, 1, 0]

    const r2 = solution24(
        ["con", "ryan"],
        ["ryan con", "ryan con", "ryan con", "ryan con"],
        3
    );
    console.log(r2); // [0, 0]
}